lib/oli/delivery/attempts/activity_lifecycle/evaluate.ex

```diff
@@ -377,44 +377,81 @@ defmodule Oli.Delivery.Attempts.ActivityLifecycle.Evaluate do
     {:ok, evaluations}
   end
 
-  @doc """
-  Evaluates an activity attempt using only the already stored state present in
-  the child part attempts.  This exists primarly to allow graded pages to
-  submit all of the contained activites when the student clicks "Submit Answers".
-  """
-  def evaluate_from_stored_input(activity_attempt_guid, datashop_session_id) do
-    part_attempts = get_latest_part_attempts(activity_attempt_guid)
-
-    # derive the part_attempts from the currently saved state that we expect
-    # to find in the part_attempts
-    part_inputs =
-      Enum.map(part_attempts, fn p ->
-        input =
-          case p.response do
-            nil -> nil
-            map -> Map.get(map, "input")
-          end
-
-        files =
-          case p.response do
-            nil -> nil
-            map -> Map.get(map, "files", [])
+  def update_part_attempts_and_get_activity_attempts(resource_attempt_id, datashop_session_id) do
+    resource_attempt_id
+    |> get_latest_activity_attempts()
+    |> Enum.reduce_while(
+      {0, [], [], []},
+      fn activity_attempt,
+         {
+           i,
+           activity_attempt_values,
+           activity_attempt_params,
+           part_attempt_guids
+         } = acc ->
+        if activity_attempt.lifecycle_state != :evaluated and activity_attempt.scoreable do
+          # --- part attempts
+          activity_attempt = activity_attempt |> Repo.preload([:revision, :resource_attempt])
+          part_attempts = get_latest_part_attempts(activity_attempt.attempt_guid)
+
+          part_inputs =
+            part_attempts
+            |> Enum.map(fn pa ->
+              {input, files} =
+                if pa.response,
+                  do: {Map.get(pa.response, "input"), Map.get(pa.response, "files", [])},
+                  else: {nil, nil}
+
+              %{
+                attempt_guid: pa.attempt_guid,
+                input: %StudentInput{input: input, files: files}
+              }
+            end)
+            |> filter_already_evaluated(part_attempts)
+
+          case activity_attempt
+               |> do_evaluate_submissions(part_inputs, part_attempts)
+               |> persist_evaluations(part_inputs, fn result -> result end, datashop_session_id) do
+            {:ok, _} ->
+              # --- activity attempt
+              part_attempts = get_latest_part_attempts(activity_attempt.attempt_guid)
```

Comment:
We are getting the part attempts again here because we need them after being updated, but if we want to avoid this select, we could try something like returning more data on the above update and work from there